src/timely-util/src/containers.rs

```diff
@@ -9,5 +9,415 @@
+                let mut alloc = Vec::with_capacity(round);
+                columnar::bytes::serialization::encode(
+                    &mut alloc,
+                    self.current.borrow().as_bytes(),
+                );
+                self.pending
+                    .push_back(Column::Align(alloc.into_boxed_slice()));
+                self.current.clear();
+            }
+        }
+    }
+
+    impl<C: Columnar> Default for ColumnBuilder<C> {
+        fn default() -> Self {
+            ColumnBuilder {
+                current: Default::default(),
+                empty: None,
+                pending: Default::default(),
+            }
+        }
+    }
+
+    impl<C: Columnar> ContainerBuilder for ColumnBuilder<C>
+    where
+        C::Container: Clone,
+    {
+        type Container = Column<C>;
+
+        #[inline]
+        fn extract(&mut self) -> Option<&mut Self::Container> {
+            if let Some(container) = self.pending.pop_front() {
+                self.empty = Some(container);
+                self.empty.as_mut()
+            } else {
+                None
+            }
+        }
+
+        #[inline]
+        fn finish(&mut self) -> Option<&mut Self::Container> {
+            if !self.current.is_empty() {
+                self.pending
+                    .push_back(Column::Typed(std::mem::take(&mut self.current)));
+            }
+            self.empty = self.pending.pop_front();
+            self.empty.as_mut()
+        }
+    }
+
+    impl<C: Columnar> LengthPreservingContainerBuilder for ColumnBuilder<C> where C::Container: Clone {}
+}
+
+/// A batcher for columnar storage.
+pub type Col2ValBatcher<K, V, T, R> = MergeBatcher<
+    Column<((K, V), T, R)>,
+    batcher::Chunker<TimelyStack<((K, V), T, R)>>,
+    ColMerger<(K, V), T, R>,
+>;
+pub type Col2KeyBatcher<K, T, R> = Col2ValBatcher<K, (), T, R>;
+
+/// An exchange function for columnar tuples of the form `((K, V), T, D)`. Rust has a hard
+/// time to figure out the lifetimes of the elements when specified as a closure, so we rather
+/// specify it as a function.
+#[inline(always)]
+pub fn columnar_exchange<K, V, T, D>(((k, _), _, _): &<((K, V), T, D) as Columnar>::Ref<'_>) -> u64
+where
+    K: Columnar,
+    for<'a> K::Ref<'a>: Hash,
+    V: Columnar,
+    D: Columnar,
+    T: Columnar,
+{
+    k.hashed()
+}
+
+/// Types for consolidating, merging, and extracting columnar update collections.
+pub mod batcher {
+    use std::collections::VecDeque;
+
+    use columnar::Columnar;
+    use differential_dataflow::difference::Semigroup;
+    use timely::container::{ContainerBuilder, PushInto};
+    use timely::Container;
+
+    use crate::containers::Column;
+
+    #[derive(Default)]
+    pub struct Chunker<C> {
+        /// Buffer into which we'll consolidate.
+        ///
+        /// Also the buffer where we'll stage responses to `extract` and `finish`.
+        /// When these calls return, the buffer is available for reuse.
+        empty: C,
```

Comment:
Why is this called `empty`? This container isn't empty most of the time, right?

Reply:
Renamed to `target`.